Per-Layer TriPlanar and Refactoring for PBR Terrain Shaders

yaRnMcDonuts opened the pull request:
Refactored AdvancedPBRTerrain and PBRTerrain to work with the new modular system, making the shader much less confusing and easier to work with.

I also added 12 new boolean matParams (1 per layer) to toggle tri-planar mode on a per layer basis. The old UseTriPlanar boolean still works to toggle triplanar for the whole terrain, but now you also can enable triplanar for  just 1 or 2 layers that need it most. (for example, mountain textures that are painted on vertical slopes typically always need triplanar to prevent noticable stretching, whereas a grass texture that is typically painted on mostly-flat areas of the terrain could get by without using tri-planar mode, or it could atleast be disabled for terrains at far distances)

Tri-planar mode has typically been the biggest reason that jme's terrains result in low FPS, so this should allow for some big optimizations when using terrains in large scenes.

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/AdvancedPBRTerrain.j3md:
```diff
@@ -169,17 +180,21 @@ MaterialDef Advanced PBR Terrain {
+ // debug the final value of the selected layer as a color output            
+        Int DebugValuesMode
             // Layers:
             //   0 - albedo (un-shaded)
```
please change 'un-shaded' to 'unshaded'

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/AdvancedPBRTerrain.j3md:
```diff
@@ -169,17 +180,21 @@ MaterialDef Advanced PBR Terrain {
             //   2 - roughness
             //   3 - metallic
             //   4 - ao
             //   5  - emissive
```
please remove the extra space before the -

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/AdvancedPBRTerrain.j3md:
```diff
@@ -279,6 +294,27 @@ MaterialDef Advanced PBR Terrain {
+            STATIC_SUN_EXPOSURE : StaticSunIntensity
+            BRIGHTEN_INDOOR_SHADOWS : BrightenIndoorShadows
+
+            USE_FIRST_LAYER_AS_TRANSPARENCY : FirstLayerAsTransparency;
```
please remove the semicolon at the end of the line

Perhaps it would be better to rename the variable 'FirstLayerAsTransparency'  to 'UseFirstLayerAsTransparency' for consistency with the other variables (e.g. UseVertexColorsAsSunIntensity and useTriPlanarMapping)

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/Modular/PBRTerrainUtils.glsllib:
```diff
@@ -0,0 +1,170 @@
+#ifndef __PBR_TERRAIN_UTILS_MODULE__
```
Please review the entire file again. Ensure the code is correctly formatted, extra whitespace is removed, and there's a single blank line between methods. Order is important when writing code; it makes it easier to review and find errors.

yaRnMcDonuts replied:
yes I'll make sure I cleanup that file before this PR is finalized.

yaRnMcDonuts commented:
@capdevon 
I believe I've addressed most of your review, let me know if I missed anything.

The only thing I didn't do yet is cleanup TerrainUtils.glsllib because I still have plans to add more functions to that file to support  PBRTerrain (right now it only works for reading texture arrays for advancedPBRTerrain).

Once everything is approved and finalized for the approach I took to refactoring AdvancedPBRTerrain, then I'll adapt things for PBRTerrain and do a final code cleanup in that file.

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/AdvancedPBRTerrain.j3md:
```diff
@@ -279,6 +294,27 @@ MaterialDef Advanced PBR Terrain {
+            STATIC_SUN_EXPOSURE : StaticSunIntensity
+            BRIGHTEN_INDOOR_SHADOWS : BrightenIndoorShadows
+
+            USE_FIRST_LAYER_AS_TRANSPARENCY : UseFirstLayerAsTransparency;
```
please remove the ; at the end of the line, after UseFirstLayerAsTransparency

capdevon commented on jme3-terrain/src/main/resources/Common/MatDefs/Terrain/Modular/AfflictionLib.glsllib:
```diff
@@ -0,0 +1,515 @@
+    void AfflictionLib_readAfflictionVector(){
+         #ifdef AFFLICTIONTEXTURE
+        
+            afflictionVector = vec4(1.0, 0.0, 1.0, 0.0); //r channel is sturation, g channel is affliction splat texture intensity, b and a unused (might use b channel for wetness eventually)
```
By 'sturation' do you by any chance mean 'saturation' ?
